modules/gapi/include/opencv2/gapi/gcomputation_async.hpp

```diff
@@ -0,0 +1,26 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
+
+#ifndef OPENCV_GAPI_GCOMPUTATION_ASYNC_HPP
+#define OPENCV_GAPI_GCOMPUTATION_ASYNC_HPP
+
+
+#include <future>
+#include <exception>                        //for std::exception_ptr
+#include <functional>                       //for std::function
+#include "opencv2/gapi/garg.hpp"            //for GRunArgs, GRunArgsP
+#include "opencv2/gapi/gcommon.hpp"         //for GCompileArgs
+
+namespace cv {
+    //fwd declaration
+    class GComputation;
+
+    GAPI_EXPORTS void                async_apply(GComputation& gcomp, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs, GCompileArgs &&args = {});
+    GAPI_EXPORTS std::future<void>   async_apply(GComputation& gcomp, GRunArgs &&ins, GRunArgsP &&outs, GCompileArgs &&args = {});
+}
```

Comment:
also `namespace gapi`?

Reply:
the problem this this approach is that ADL will not work then (as ```async``` and ```GCompiled```) will be in different namespaces. i.e. following code will not compile without either full qualification of the function (```cv::gapi::async_apply```) or ```using```  directive/clause: 

```c++
cv::GMat in;  
cv::GMat  out = in + in;
cv::GCompuation g{in,out};
cv::Mat m, r;
//here will be a compile error
// async_apply(g, cv::gin(m), cv::gout(r));

// use it either like this:
cv::gapi::async_apply(g, cv::gin(m), cv::gout(r));

// or like this:
using cv::gapi::async_apply;

async_apply (g, cv::gin(m), cv::gout(r));
```

maybe add **g** to name - i.e. ```gasync_apply```?

Reply:
fixed